[Clock] A new component to decouple applications from the system clock

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

After watching @afilina's talk at SymfonyWorld Online last week + listening to her using a "clock" service to improve testability of time-sensitive logics, I decided to propose this new "Clock" component. This also relates to the ongoing efforts to standardize a `ClockInterface` in [PSR-20](https://github.com/php-fig/clock).

This PR provides a `ClockInterface`, with 3 methods:
- `now(): \DateTimeImmutable` is designed to be compatible with the envisioned PSR-20;
- `sleep(float|int $seconds): void` advances the clock by the provided number of seconds;
- `withTimeZone(): static` changes the time zone returned by `now()`.

The `sleep()` methods takes inspiration from `ClockMock` in the PhpUnitBridge, where this proved useful to improve testability. Ideally, we could use this component everywhere measuring the current time is needed and stop relying on  `ClockMock`.

This PR provides 3 clock implementations:
- `NativeClock` which relies on the system clock;
- `MockClock` which allows mocking the time;
- `MonotonicClock` which relies on `hrtime()` to provide a monotonic clock.

If this gets accepted, I'll follow up with a PR to add clock services to FrameworkBundle and we'll then be able to see where we could use such clock services in other components. I hope PSR-20 will be stabilized by Symfony 6.2, but this is not required for this PR to be useful.

nicolas-grekas committed:
[Clock] A new component to decouple applications from the system clock

theofidry commented on src/Symfony/Component/Clock/MockClock.php:
```diff
@@ -0,0 +1,77 @@
+{
+    private \DateTimeImmutable $now;
+
+    public function __construct(\DateTimeImmutable|\DateTimeZone|string $now = null)
```
In my experience a test clock implementation benefits from having a `freeze(DateTimeImmutable): void` method. This is more useful for integration tests where you might need to set the clock at a specific time or different times in a row.

Also probably out of scope but I like to pair it up with a PHPUnit extension which unfreeze it between each test (a bit similar to https://github.com/dmaicher/doctrine-test-bundle)

nicolas-grekas replied:
A MockClock is always frozen so I'm not sure what these would do. To advance the clock, one can call `$mockClock->sleep(123.456)`, eg between test cases with a phpunit extension, or between the steps of a test case when needed.

theofidry replied:
Ah missed that part: yes I went the other road of having a `FreezableClock`, which is not frozen by default, but can be arbitrarily frozen. So it is probably better to go with a dedicated class instead

drupol replied:
I have the feeling that it's indeed better to have its own class and keep them immutable.

drupol reviewed (approved):
Cool initiative!

heiglandreas commented on src/Symfony/Component/Clock/HrClock.php:
```diff
@@ -0,0 +1,69 @@
+        $hrtime = hrtime();
+        $now = '@'.$hrtime[0].'.'.substr(str_pad($hrtime[1], 9, 0, \STR_PAD_LEFT), 0, -3);
+
+        return (new \DateTimeImmutable($now, $this->timezone))->setTimezone($this->timezone);
```
Setting the timezone in the constructor can be omitted as the object is instantiated using a timestamp - which by definition contains a timezone-declaration so that the second parameter to the constructor is omitted. (see [documentation note](https://www.php.net/manual/en/datetimeimmutable.construct.php#refsect1-datetimeimmutable.construct-parameters))

In addition I'm not sure the extra calls around `hrtime` make the DateTimeImmutable object more precice than calling `return (new DateTimeImmutable())->setTimezone($this->timezone);`

And I'm open for debate whether the `setTimezone()` is necessary at all. But that might be a philosophy topic. Modifying the point of observation of a point in time is IMO not something the Clock should have to take care of.

heiglandreas replied:
Apart from that I'm getting interesting results from that when testing it in PHP: https://3v4l.org/G4jtQ

And the advantage of the HighResolution timing somehow is eliminated by reducing the nanoseconds to microseconds. As there seems to be no added precision I'd question whether this class is necessary at all...

nicolas-grekas replied:
> Setting the timezone in the constructor can be omitted as the object is instantiated using a timestamp - which by definition contains a timezone-declaration so that the second parameter to the constructor is omitted

True, yet it's marginally but still measurably faster to give this argument, that's why I kept it :shrug: 

> In addition I'm not sure the extra calls around hrtime make the DateTimeImmutable object more precice than calling return (new DateTimeImmutable())->setTimezone($this->timezone);

The call to hrtime is not meant to make the timestamp more precise, but to ensure that `now()` and `time*()` methods return consistent monotonic results.

> And I'm open for debate whether the setTimezone() is necessary at all. But that might be a philosophy topic. Modifying the point of observation of a point in time is IMO not something the Clock should have to take care of.

That's a constructor argument so it's implementation specific. I think some ppl will need to set the TZ of all of their DateTimeImmutable objects so I prefer giving control over the TZ to all implementations.

> Apart from that I'm getting interesting results from that when testing it in PHP: [3v4l.org/G4jtQ](https://3v4l.org/G4jtQ)

Yep :)  hrtime() counts "from an arbitrary point in time". Still suitable for relative comparisons.

> And the advantage of the HighResolution timing somehow is eliminated by reducing the nanoseconds to microseconds. As there seems to be no added precision I'd question whether this class is necessary at all...

The other advantage is worth it: monotonicity.
And lets' not forget about the `time*()` methods when one needs sub-µs accuracy.

heiglandreas commented on src/Symfony/Component/Clock/HrClock.php:
```diff
@@ -0,0 +1,69 @@
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class HrClock implements TimeInterface
```
I'd add the ClockInterface and a possible SleepInterface as well here to separate the different concerns better.

* the ClockInterface to get the current Point in Time (which can then be mocked to represent any point in time)
* the TimeInterface for a convenient way of getting the time including microseconds (which could perhaps also contain a `timeInt` method to get the timestamp as integer),
* the SleepInterface for a convenient way to control the flow of the application. As we have discussed regarding PSR-20 `sleep` has nothing to do with getting a point of time but with controlling the flow of the application.

nicolas-grekas replied:
I don't think splitting in many interfaces would provide any benefits: all those methods are tightly related and fit a single abstraction to me. Having a single `TimeInterface` also provides better DX, with no drawbacks IMHO.

> which could perhaps also contain a timeInt method to get the timestamp as integer

Good idea, PR updated.

heiglandreas replied:
> I don't think splitting in many interfaces would provide any benefits: all those methods are tightly related and fit a single abstraction to me. Having a single `TimeInterface` also provides better DX, with no drawbacks IMHO.

To me they are three completely separate use-cases:

* Getting a dedicated Point in the dimension of time. (`now`)
* getting a number of seconds after a specific point in time (`timeArray` and `timeFloat`)
* stopping the execution of the script for a dedicated amount of time.

Especially the last one seems to be something unrelated to timekeeping in terms of *the interface*. It might absolutely make sense to have them within *the implementation*  IMO.

But that feels like a debate of principles and in the end I'm not the one taking that decision for symfony (luckily :wink:).

nicolas-grekas replied:
> Especially the last one seems to be something unrelated to timekeeping in terms of the interface. It might absolutely make sense to have them within the implementation IMO.

I don't have the same view on that topic: to me, sleep() is like a setter and the other methods are sort of getters. Segregating setters from getters doesn't make sense when they work on the same state.

About segregating `now` and `time*` I'd prefer not because it makes using the abstraction more complex for little to no practical benefits. Providing them at the abstraction level on the other end ensures that all those methods are always provided by implementations, enforcing the synchronization of the returned time by all methods.

GromNaN replied:
The method `sleep` could be decorated with a stopwatch so that sleeping time can be visualized in the profiler. I would say that having less methods in the interface is better for decoration.

nicolas-grekas replied:
Marking as resolved after https://github.com/symfony/symfony/pull/46715#issuecomment-1165415212

heiglandreas commented on src/Symfony/Component/Clock/MockClock.php:
```diff
@@ -0,0 +1,77 @@
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class MockClock implements TimeInterface
```
Instead of calling the class by what it should be used for it might make more sense to call the class by what it actually does. In this case that would be a `FrozenClock` which freezes the point in time to a specific one. 

Or to multiple specific ones when the constructor also accepts an array of DateTimeImmutables.

A possible further Implementation might then be a `FreezableClock` that can be frozen multiple times using @theofidry's idea via a `freeze` method.

nicolas-grekas replied:
> Instead of calling the class by what it should be used for it might make more sense to call the class by what it actually does. In this case that would be a FrozenClock which freezes the point in time to a specific one.

I prefer keeping MockClock. I think the name kinda says both what it might be use for *and* what it does, thus providing better discoverability (and is consistent with "MockHttpClient".)

> Or to multiple specific ones when the constructor also accepts an array of DateTimeImmutables.

`now()` would then return dates from that sequence? Interesting idea :)

> A possible further Implementation might then be a FreezableClock that can be frozen multiple times using @theofidry's idea via a freeze method.

Agreed. I might wait for someone that needs that to submit a PR instead of providing it in this initial PR though.

eerison replied:
Hey @nicolas-grekas 

I saw that your first `MockClock` version wasn't `final`, But I can see that all class are final, what made you add classes as final?

I ask this because I see others symfony's classes implementing an interface But those aren't final.
for example: https://github.com/symfony/serializer/blob/7.0/Encoder/JsonEncode.php#L21

 is there any reason to make those clock classes as final and in others repositories not?

nicolas-grekas replied:
It's subtle and certainly never a black and white decision. Why are you asking?

eerison replied:
Because I see some open source projects closing their classes for extensions and force the dev implement the interface, or maybe decarorate the classes in case they want to reuse some existing code.  And as far as I know it is to help maintainers handle BC easily, because the dev won't extend classes. 


But why symfony doesn't close the api and make the dev always implement the interface?

And as this clock classes are different of others,  I thought it has a different reason for this. 

BTW, thank you for your time in replying me, I really appreciate that.

nicolas-grekas replied:
In Symfony we're not dogmatic on the topic. I think we learned that from a maintenance perspective, making classes final makes it easier to evolve them without too much BC concerns. But our core desire is to empower end users, and "final" comes into the way sometimes so we're open to reconsidering. As for why some classes are not final: history for some, and openness to inheritance for others.
The open/closed principle is a balanced one.

eerison replied:
So nice ♥️ 

> But our core desire is to empower end users.

Do you mean give them more flexibility?

nicolas-grekas replied:
Yes, I mean more flexibility to achieve what they need to achieve, and some framework to guide them in the process :)

eerison replied:
This vision is so nice,  because even you know that closes classes are easier for maintaining, but symfony prefer give flexibility for users,  even if it costs more work from symfony side 🤯♥️

Well thank you again for your time

heiglandreas reviewed (commented):
I doubt that PSR-20 is going to be ready by the release of Symfony 6.2. Though in the meantime there is the forward-compatible [stella-maris/clock](https://packagist.org/packages/stella-maris/clock) interface that can be used and already is implemented by some of the most used clock-implementations. 

With that in mind I'd think about - for the sake of interface segregation - removing the `now` function completely from the `TimeInterface` (and perhaps also moving the `sleep` function into a separate interface - we've discussed that during the PSR-20 discussions IIRC). The different implementations can then still implement the different interfaces in one class. And should more than one interface be required in a method we can by now require an implementation via a UnionType.

drupol commented:
> With that in mind I'd think about - for the sake of interface segregation - removing the `now` function completely from the `TimeInterface`. The different implementations can then still implement the different interfaces in one class. And should more than one interface be required in a method we can by now require an implementation via a UnionType.

Hi Andreas,

Why would we want to remove the `now` method? The whole point of this component is actually to have such method.
Could you please develop?

nicolas-grekas commented:
> I doubt that PSR-20 is going to be ready by the release of Symfony 6.2.

:shrug: hurry up PHP-FIG :)

> for the sake of interface segregation - removing the now function completely from the TimeInterface (and perhaps also moving the sleep function into a separate interface - we've discussed that during the PSR-20 discussions IIRC).

I'd better not as I explained in https://github.com/symfony/symfony/pull/46715#discussion_r901352375
If we end up being incompatible with PSR-20, we'll deal with it with a deprecation layer.
But looking at the state of PSR-20, it looks unlikely that we'll have to do that.

nicolas-grekas commented:
Thanks for the review btw @heiglandreas !

heiglandreas commented:
> > With that in mind I'd think about - for the sake of interface segregation - removing the `now` function completely from the `TimeInterface`. The different implementations can then still implement the different interfaces in one class. And should more than one interface be required in a method we can by now require an implementation via a UnionType.
> 
> Hi Andreas,
> 
> Why would we want to remove the `now` method? The whole point of this component is actually to have such method. Could you please develop?

The `now`-method would be already defined in a `ClockInterface` - like `psr/clock` or `stella-maris/clock`, so defining it here again in a `TimeInterface` is not really necessary. The actual *Implementation* then would implement `TimeInterface, ClockInterface, SleepInterface`.

So the component would either require an external `ClockInterface` or declare a spearate `ClockInterface` in parallel to the current `TimeInterface` and a currently not existing `SleepInterface`.

bendavies commented on src/Symfony/Component/Clock/MockClock.php:
```diff
@@ -0,0 +1,91 @@
+namespace Symfony\Component\Clock;
+
+/**
+ * A clock that returns always the same date, suitable for testing time-sensitive logic.
```
just a native english suggestion
```suggestion
 * A clock that always returns the same date, suitable for testing time-sensitive logic.
```

nicolas-grekas replied:
thanks Ben, fixed

bendavies commented on src/Symfony/Component/Clock/HrClock.php:
```diff
@@ -0,0 +1,83 @@
+namespace Symfony\Component\Clock;
+
+/**
+ * A high-resolution monotonic clock, counted from an arbitrary point in time.
```
```suggestion
 * A high-resolution monotonic clock, counting from an arbitrary point in time.
```

nicolas-grekas commented:
> The now-method would be already defined in a ClockInterface - like psr/clock or stella-maris/clock, so defining it here again in a TimeInterface is not really necessary. The actual Implementation then would implement TimeInterface, ClockInterface, SleepInterface.

I want the component to be standalone and complete on its own. When PSR-20 will be ready, we might extend it, but I wouldn't make a stable PSR-20 a requirement to this component.

> So the component would either require an external ClockInterface or declare a spearate ClockInterface in parallel to the current TimeInterface and a currently not existing SleepInterface.

Replying in the main thread to help readers follow the discussion: see https://github.com/symfony/symfony/pull/46715#discussion_r901498040 about this topic.

HeahDude commented on src/Symfony/Component/Clock/HrClock.php:
```diff
@@ -0,0 +1,83 @@
+     */
+    public function timeFloat(): float
+    {
+        return hrtime(true) / 1E9;
```
What about using named arguments `hrtime(as_number: true)`?

nicolas-grekas replied:
I don't think that's needed. The doc is here if ppl wonder about the arguments of a function.

HeahDude commented on src/Symfony/Component/Clock/TimeInterface.php:
```diff
@@ -0,0 +1,39 @@
+{
+    public function now(): \DateTimeImmutable;
+
+    public function sleep(float $seconds): void;
```
`int|float`?

If an end-user class using strict types calls this API, could it be a problem?
(https://3v4l.org/mst1W, `sleep` would be called in context where strict types in not defined).
Implementations may need explicit cast, maybe some tests are missing?

nicolas-grekas replied:
strict mode :facepalm:
But anyway, good idea, PR updated (but I won't add tests specific to this, testing the engine is not needed.)

chalasr commented on src/Symfony/Component/Clock/README.md:
```diff
@@ -0,0 +1,13 @@
+Clock Component
+===============
+
+Symfony Clock enables decoupling applications from the system clock.
```
we add some code snippet to the README for new components, replacing the component docs IIRC

nicolas-grekas replied:
README updated

nicolas-grekas replied:
README updated

afilina commented on src/Symfony/Component/Clock/HrClock.php:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function timeArray(): array
```
Not a fan of using arrays, less so in interfaces.

Besides, I'm not sure why this interface needs anything beyond `now`. Everything else looks like it doesn't belong here. This class is mixing too many responsibilities. I think there should be a difference between a Clock interface, whose job is merely to give the current time, and a component for all things time-related.

It's better to start small and add/compose only if absolutely needed. Splitting an interface later will be a nightmare.

nicolas-grekas replied:
Hi Anna, thanks for chiming in.

> Not a fan of using arrays, less so in interfaces.

Yep, yet it's the common way to return a tuple of numbers when the size of integers is too small to properly count the time, typically on 32-bit arch. I don't think adding a VO for that is worth it.

>  I'm not sure why this interface needs anything beyond now

The plan is to have PSR-20 be the super simple interface that fits most use cases.

`TimeInterface` will extend PSR-20 once it's available and will be useful when performance matters (using DateTimeImmutable creates overhead that isn't desired when eg writting a StopWatch) or when needing to abstract sleep().

Of course, until PSR-20 is available, ppl won't have a better choice than using `TimeInterface`. I considered adding a `ClockInterface` that'd contain only `now()` but I'd better not since PSR-20 might just be the one. IF it gets released soon enough of course. I'm fine adding `ClockInterface` here also but we're going to have a hard time deprecating it when PSR-20 will come. I'd rather encourage the PHP-FIG to release PSR-20 by November...

heiglandreas replied:
As already mentioned: there is an interface that is used by the most-used Clock-Libraries that is also forward-compatible to the current draft of PSR-20 that could be used *right away*

nicolas-grekas replied:
> As already mentioned: there is an interface that is used by the most-used Clock-Libraries that is also forward-compatible to the current draft of PSR-20 that could be used right away

Sorry I didn't comment on this topic: we never referenced any "interop" packages as that doesn't fit out maintenance policies, per the package principles. We can already anticipate that referencing [stella-maris/clock](https://packagist.org/packages/stella-maris/clock) would come at a price that the community will have to pay when we'll replace it with PSR-20. Instead, I prefer have the community use `TimeInterface` even if it does a bit too much for the common use case: this won't incur any "forced" migration costs and it will be easy to migrate to PSR-20 when possible.

heiglandreas replied:
The `stella-maris/clock` package will extend `psr/clock` once it is released (and based on the current draft). So all packages implementing or extending that interface will immediately implement or extend `psr/clock`. And it will stay as long as necessary so that the community will not have to pay any price at all when it will be replaced by PSR-20. Provided that PSR-20 will be released as it is currently drafted.

Just as a background-information.

nicolas-grekas replied:
Thank you for that!
But unfortunately that doesn't solve the issue for Symfony: if we build on the `StellaMaris` namespace, we will have a hard time removing it without breaking BC. Removing a super type is generally very hard thus costly for the community.

heiglandreas replied:
Though you wouldn't *need* to remove it. It will either extend the (hopefully soon) to be coming PSR-20 should it be accepted as it is currently drafted (then just being an empty shell that doesn't *need* to be replaced but could over time). Or - in the event of PSR-20 *not* being accepted as it is - it will stay like this ... forever is kind of a really long time, but let's just say it will stay as long as necessary and in a sustainable way.

So no *need* to remove it and therefore not costly for the community.

afilina replied:
I think that this is suffering from scope creep. The initial problem is "how do we unit-test code that relies on the current date?" A simple interface with `now` solves it quickly and painlessly, in a dozen lines of code. The need to conform to some other interface is unnecessary in this context. You can always create a bigger component later (if you have a problem that needs solving), in which you can then reuse the `DateTimeClock`. You don't need to figure out the entire roadmap of time handling to solve this one small problem.

This is from my [blog](https://afilina.com/misc-util-functions). That's all you really need (exact implementation up to you):
```php
interface Clock
{
    public function now(): DateTimeImmutable;
}

final class DateTimeClock implements Clock
{
    public function now(): DateTimeImmutable
    {
        return new DateTimeImmutable(
            'now', 
            new DateTimeZone('UTC')
        );
    }
}
```

nicolas-grekas replied:
Marking as resolved after https://github.com/symfony/symfony/pull/46715#issuecomment-1165415212

stof commented on src/Symfony/Component/Clock/MockClock.php:
```diff
@@ -0,0 +1,91 @@
+     */
+    public function now(): \DateTimeImmutable
+    {
+        return clone $this->now;
```
no need to clone a DateTimeImmutable, as it is immutable.

nicolas-grekas replied:
It's an implementation detail but I did it his way to not leak the fact that the same time is always returned at the identity level.
This is tested by `$this->assertNotSame($clock->now(), $clock->now());`

chalasr reviewed (commented):
Cool

stof commented on src/Symfony/Component/Clock/TimeInterface.php:
```diff
@@ -0,0 +1,39 @@
+    public function timeFloat(): float;
+
+    /**
+     * Gets the current unix timestamp.
```
This description is an issue, because HrClock does not return the unix timestamp (as it is not based on the unix EPOCH)

nicolas-grekas replied:
I removed the word "unix", I hope that's enough.

stof commented on src/Symfony/Component/Clock/composer.json:
```diff
@@ -0,0 +1,27 @@
+{
+    "name": "symfony/clock",
+    "type": "library",
```
I suggest adding the `clock` keyword too

nicolas-grekas replied:
good idea, updated

chalasr commented on src/Symfony/Component/Clock/README.md:
```diff
@@ -0,0 +1,48 @@
+use Symfony\Component\Clock\NativeClock;
+use Symfony\Component\Clock\TimeInterface;
+
+class MyClockSensiticeClass
```
s/Sensitice/Sensitive I guess, same l37 below

nicolas-grekas replied:
fixed thanks

lyrixx commented on src/Symfony/Component/Clock/MockClock.php:
```diff
@@ -0,0 +1,91 @@
+{
+    private \DateTimeImmutable $now;
+
+    public function __construct(\DateTimeImmutable|\DateTimeZone|string $now = null)
```
Why don't you add another argument for the timezone? It would make the API cleaner & and this code simpler too.
Especially since we have named argument.

nicolas-grekas replied:
I thought about it but it looks like those two arguments would be mutually exclusive. Passing both a TZ and a DateTime wouldn't make sense.

jderusse replied:
but passing a **string** and a TZ make sense. 
use cases:
- User input in console / or form with `type=text`
- SGDB with poor date type like MySQL

I'm fine with `__construct(\DateTimeImmutable|\DateTimeZone $now = null)`

GromNaN replied:
Having an argument name that matches the expected value is more important than setting `null` in an argument. Having 2 arguments can be useful if the date part does not specify a timezone.

```suggestion
    public function __construct(\DateTimeImmutable|string $now = null, \DateTimeZone $timezone = null)
```

Usage:
```php
new MockClock('2022-06-20 18:20:00', 'Europe/Paris');
new MockClock(null, 'Europe/Paris');
```

nicolas-grekas replied:
> but passing a string and a TZ make sense.

That convinced me :)
PR updated to split this in two arguments.

javiereguiluz commented:
Thanks for this proposal!

In the description we see this:

```
timeInt(): float returns the current timestamp;
```

I guess it's a typo that `timeInt()` returns a `float` instead of `int`.

- - - - -

If the naming of the class/methods is up for debate, here's a proposal for your consideration:

```php
// Before
interface TimeInterface
{
    public function now(): \DateTimeImmutable;
    public function sleep(float|int $seconds): void;
    public function timeInt(): int;
    public function timeFloat(): float;
    public function timeArray(): array;
}

// After
interface ClockInterface
{
    public function now(): \DateTimeImmutable;
    public function sleep(float|int $seconds): void;
    public function timestamp(): float;
}
```

I'd remove `timeInt()`, `timeFloat()` and `timeArray()` ... but you said that they are important for performance reasons. So, why not merging all of them into a single method called `timestamp()` which returns a timestamp with microseconds precision. It'd be equivalent to `timeFloat()` ... and the `timeInt()` value can be obtained as `(int) $clock->timestamp()`. About the `timeArray()` method, I don't understand well its need.

Thanks.

nicolas-grekas commented:
> I guess it's a typo that timeInt() returns a float instead of int.

Good catch, PR description updated.

> // After
> interface ClockInterface

We'd rather use another name than the one planned for PSR-20. This will reduce confusion.

> I'd remove timeInt(), timeFloat() and timeArray()

We cannot remove `timeArray()` because that would mean making it impossible to have decently accurate timestamps on 32-bit.

We could remove `timeInt()`. But that would force writing an explicit cast for ppl that use strict types. Consider this:
`date('Y-m-d', (int) $clock->time())`
vs
`date('Y-m-d', $clock->timeInt())`

Note that on 32-bit archs, `(int) $clock->time())` is less accurate than `$clock->timeInt()`. Being accurate on 32-bit archs would require using `$clock->timeArray()[0]`. That might be a good reason to keep the `timeInt()` method.

I realized that we could rename `timeFloat()` to just `time()`, so I just did. Makes sense ?

derrabus commented on src/Symfony/Component/Clock/MockClock.php:
```diff
@@ -0,0 +1,91 @@
+    public function __construct(\DateTimeImmutable|\DateTimeZone|string $now = null)
+    {
+        if (\is_string($now ??= 'UTC')) {
+            try {
+                $now = new \DateTimeZone($now);
+            } catch (\Exception) {
+                $now = new \DateTimeImmutable($now);
+            }
```
```suggestion
            $now = new \DateTimeImmutable($now);
```

A timezone identifier is valid input for `DateTimeImmutable`'s constructor.

Those two calls are equivalent:
```php
new DateTimeImmutable('now', new DateTimeZone('Europe/Berlin'));
new DateTimeImmutable('Europe/Berlin');
```

nicolas-grekas replied:
> A timezone identifier is valid input for DateTimeImmutable's constructor.

Wow, I didn't know that :)
PR updated to handle two arguments.

derrabus commented:
First of all: Thank you for your proposal! I do believe that such a component is a great addition to Symfony. And even better if such a component would allow us to move `ClockMock` out of PhpUnitBridge eventually.

Regarding the `HrClock`, I don't think that it is particularly useful in its current form. The numbers that we get from `hrtime()` are arbitrary and only useful when calculating elapsed time. However, you treat them as Unix timestamps which yields funny dates somewhere in the 70's or 80's depending on where you run your code.

However, I believe that for certain tests it is useful to have a clock that I can set to a certain point in time without freezing it. My proposal is a `HrClock` that is initialized with a `DateTimeImmutable` instance, the same way `MockClock` is. But whenever that clock is queried for a current time, we use `hrtime()` to calculate an offset that is added to that initial time.

GromNaN reviewed (commented):
I love this new component. It allows dependency injection for something that is usually a global state. I have a similar class in my projects and this is very handy to be able to freeze the time for functional tests.

Do you plan other implementations of `sleep()` for async frameworks? [AMP's `delay` function](https://github.com/amphp/amp/blob/v3/examples/async.php#L10) and ReactPHP's [Timer `sleep`](https://reactphp.org/promise-timer/#sleep). Thanks to fibers, the interface for sync and async will be the same.

This interface was suggested in https://github.com/symfony/symfony/issues/41520#issuecomment-853331607.

> I'll follow up with a PR to add clock services to FrameworkBundle.

We will need multiple clock services (example: UTC for DB queries and Local TZ for display).

nicolas-grekas commented:
> My proposal is a HrClock that is with an initial DateTimeImmutable instance

Would be nice, but there is no way to implement this: `hrtime()` cannot be seeded and simulating this with `microtime()` would reduce accuracy and break monotonicity.

> Do you plan other implementations of sleep() for async frameworks?

I don't but you know how this works (PR welcome ;) )

> We will need multiple clock services (example: UTC for DB queries and Local TZ for display).

I think so yes!

nicolas-grekas commented:
> But whenever that clock is queried for a current time, we use hrtime() to calculate an offset that is added to that initial time.

Hum, OK, but then we might add a significant overhead. Could be worth a try, in a follow up PR ideally.

theofidry commented:
@nicolas-grekas I suppose you already have a few strong use cases for `HrClock`? Might be worth sharing them as I doubt many people make use of `hrtime()` and so some reasons as to why you went ahead with this and some decision you took their could be better understood with additional context.

Regarding splitting the interfaces (`SleepInterface`, `ClockInterface`, `TimeInterface`), I think the question is more about the usages. As the cost is:

- if they are too split, you'll often find yourself having to typehint not just one but several interfaces
- if they are not split enough, implementing your own is more tedious

I admit in practice I never needed more than `ClockInterface` and a system + freezable implementation so it's hard to tell if you really need `timeFloat()` & `timeArray()` when you're relying on `sleep()` for example.

nicolas-grekas commented:
> I suppose you already have a few strong use cases for HrClock? Might be worth sharing them as I doubt many people make use of hrtime() and so some reasons as to why you went ahead with this and some decision you took their could be better understood with additional context.

I don't have a strong case. I know what `hrtime()` is for: profiling things. So that's the use case for `HrClock` also I guess. Not many ppl will use it I agree, but that's still a valid clock implementation, with its unique characteristics.

> it's hard to tell if you really need timeFloat() & timeArray() when you're relying on sleep() for example.

I checked the code base of Symfony: next to almost every call to `sleep()` there is a call to `microtime()`.

matthiasnoback commented:
I think `sleep()` isn't a good name, because it sounds like PHP's `sleep()` function which would actually sleep/halt the execution, whereas if I understand correctly, using it on the clock means only moving the time forward. Also, I'd personally be looking for an actual "setter", not just move the time forward by a number of seconds, but going directly to a specific moment.

Where you inject the clock interface, you wouldn't need access to setter functions like `sleep()` anyway. In fact, you'd be mistaken to use this function. Only a test would want to call this function, so to me it would make sense to not have this method on the interface. Maybe not even on any interface, only on the concrete mock clock. Throwing around principles doesn't always make sense, but in this case consider Liskov; not every clock implementation can provide an implementation for `sleep()` anyway.

derrabus commented:
> I think `sleep()` isn't a good name, because it sounds like PHP's `sleep()` function which would actually sleep/halt the execution

If you call `sleep()` on the `NativeClock`, this is exactly what happens.

nicolas-grekas commented:
@matthiasnoback thanks for checking this PR. I think you might have misunderstood the purpose of the sleep() method in the interface. It is *really* aimed at implementations, to allow them to abstract out a way to pause the flow of some logic. Turning sleep into a setter could be useful only for tests, but this is not the only use case that this targets.

heiglandreas commented:
> @matthiasnoback thanks for checking this PR. I think you might have misunderstood the purpose of the sleep() method in the interface. It is _really_ aimed at implementations, to allow them to abstract out a way to pause the flow of some logic. Turning sleep into a setter could be useful only for tests, but this is not the only use case that this targets.

But when `sleep()` is about execution control, then why is that the responsibility of a clock? See [this previous discussion](https://github.com/symfony/symfony/pull/46715#discussion_r901285420)

stof commented:
@heiglandreas the point of `sleep` is that this method will return once the clock has slept for the given time. For a native clock, that means using `sleep()`. For a mock clock, that can instead be implemented by increasing the fixed time of that clock and returning immediately (giving you fast tests). That's why it is a concern related to the clock implementation.

heiglandreas commented:
> @heiglandreas the point of `sleep` is that this method will return once the clock has slept for the given time. For a native clock, that means using `sleep()`. For a mock clock, that can instead be implemented by increasing the fixed time of that clock and returning immediately (giving you fast tests). That's why it is a concern related to the clock implementation.

A clock never sleeps! You may halt the execution of an *application* but that has nothing to do with a clock. Time passes no matter whether the application is halted or not. And the only responsibility of a clock is to return the current point in time (mocked or not). 

When you want to stop the application for a given amount of time and you need the new time afterwards, then you call the clock and it will happily provide you with the new time.

stof commented:
@heiglandreas but then, how would you mock the sleeping in MockClock if you don't interact with it ?

heiglandreas commented:
As @matthiasnoback recommended: The MockClock can have a method that handles that. Or interacts with the call to `sleep`.

But that's an implementation detail unrelated to the interface.

stof commented:
@heiglandreas but if the code calls the native `sleep()` because there is no such method on the clock, then there is no way to replace that with advancing the mocked clock instead of making the testsuite actually wait for 10s.

heiglandreas commented:
That's why I recommended a `SleepInterface` (unrelated to the clock) that can be used to mock the sleep. Alternative would be something that I right now have in my head that I'll code out later today.

ro0NL commented:
> That's why I recommended a SleepInterface (unrelated to the clock) that can be used to mock the sleep

it should be noted sleep() is related to "TimeInterface", not "ClockInterface"

a mocked sleep can never forward a native clock, which proves things are related

at this point im wondering if the PSR-20 design is flawed/insufficient :thinking:

heiglandreas commented:
Regarding the "never

> > That's why I recommended a SleepInterface (unrelated to the clock) that can be used to mock the sleep
> 
> it should be noted sleep() is related to "TimeInterface", not "ClockInterface"
> 
> a mocked sleep can never forward a native clock, which proves things are related 

That depends on how the MockedClock is implemented. I've roughly drafted two possiblilities [in this gist](https://gist.github.com/heiglandreas/70979dbf402e1983448f134b9c96093c) and I'm sure there are many others that people will come up with. And all of them will be SOLID

> 
> at this point im wondering if the PSR-20 design is flawed/insufficient thinking

The PSR-20 design was chosen deliberately that way. The sleep-functionality was a big discussion point and it was decided that a function for flow-control is not in the responsibility of a clock.

ro0NL commented:
a _mocked sleep_ can never forward a _native clock_ :)

> You may halt the execution of an application but that has nothing to do with a clock. 

halting affects the clock 100% ;)

in real life i can fast forward a clock, i cannot fast forward Coordinated Universal Time

following that logic, i think PSR designed TimeInterface, whereas Symfony actually designs ClockInterface

ro0NL commented on src/Symfony/Component/Clock/TimeInterface.php:
```diff
@@ -0,0 +1,39 @@
+ */
+interface TimeInterface
+{
+    public function now(): \DateTimeImmutable;
```
WDYT about `now(?DateTimeZone $tz  = null)` to bypass TZ conversion and being more consise also

```
$clock->now($user->getTz())
$clock->now()->setTz($user->getTz())
```

ro0NL replied:
i wouldnt mind an accessor for default TZ either :)

nicolas-grekas replied:
The accessor and the setter already exist: `->now()->get/setTimezone()`.
If you mean changing the default TZ of the clock service, we could consider adding `->withTimeZone()`.
I'd just prefer postponing the discussion on this idea to a follow up PR.

ro0NL replied:
i meant "to bypass TZ conversion and being more consise also"

so similar like we do `time()` over `now()->getTimestamp()`, isnt it?

i actually prefer the shorthands, esp. `now($tz)`

nicolas-grekas replied:
What would be the benefit of this bypass? I doubt this would make any difference if you mean perf. If not perf, then I prefer a simple yet capable API.

ro0NL replied:
i should have measured :} i see some gain, but i think the numbers are negligible yes

i was mostly pursuing now($tz) for more consise API, it's not a blocker

matthiasnoback commented:
Thanks for explaining the purpose of `sleep()` @nicolas-grekas. I agree then with @afilina that it seems like things are added to the clock that don't really need to be there. Of course, for the purpose of using these interfaces in Symfony you should do whatever makes sense ;) But from the perspective of user-developed code, the idea of a clock is just that you don't have `new DateTimeImmutable('now')` all over the code base, which makes testing any of it hard. That's why I wouldn't turn it into a component. Every new project can just have that one interface and those two really small implementations that @afilina shows. That's exactly what I've always done so far.

theofidry commented:
@matthiasnoback but in that case people can just type-hint `ClockInterface` right? The extra methods you are referring to are from `TimeInterface` which is more rich and for a slightly different purpose

nicolas-grekas commented:
Brain activity at night made me realize that we could remove the `timeInt()` and `timeArray()` variants and keep only `time()`, provided `time()` returns a string. As a reminder, the concern this addresses is loss-less time measurements on 32-bit archs. This is the same concern that leads to having `microtime(false)` also return a string.

Then I did my homework and tried to verify my own claim about performance, which is the justification I've provided for these `time*()` methods.

Comparing `new \DateTimeImmutable('now', $tz)` to `microtime(true)`, the latter is almost 5x faster.
But when comparing `new \DateTimeImmutable('now', $tz)` to `(string) microtime(true)`, the latter is only 1.1x faster!

This difference is not significant enough to warrant my claim. I'm thus removing all `time*()` methods from my proposal. Performance-critical time measurements might continue to use `ClockMock` since we can't abstract time measurements without adding overhead to it. Note that using the component for not-so-critical performance measurements is still valid, because we're talking about quite fast measurements.

Since this means the interface now contains only `now()` and `sleep()`, I've renamed it to `SleepableClockInterface`.

I've also implemented the offset logic @derrabus suggested for `HrClock`, which is now renamed `MonotonicClock` (because it's not "high-resolution" anymore, see notes about the overhead above.)

I updated the PR and the description above with all those changes.

I think this addresses most if not all concerns raised so far.

The last thing that remains is the lack of a `ClockInterface`, and that's on the PHP-FIG :pray:

lchrusciel commented:
I'm looking forward to the newest PSR-20 and this component! At Sylius, we've recently introduced the [Calendar component](https://github.com/Sylius/Calendar), but we will surely try to migrate to the use yours. Our use case mainly decouples implementation from `new` in the code and provides ease-of dates testing - we need to freeze the clock and set its time for some given moment. If you are open to SettableClock implementation and interface, I would love to try it.

nicolas-grekas commented:
> we need to freeze the clock and set its time for some given moment. If you are open to SettableClock implementation and interface, I would love to try it.

Sure we are if this proves needed :)

derrabus commented on src/Symfony/Component/Clock/MonotonicClock.php:
```diff
@@ -0,0 +1,73 @@
+final class MonotonicClock implements SleepableClockInterface
+{
+    private int $sOffset;
+    private int $µsOffset;
```
As someone who doesn't have Greek letters on his keyboard, I'd like to propose sticking to ASCII characters for variable names. 😓

nicolas-grekas replied:
granted :)

derrabus commented on src/Symfony/Component/Clock/Tests/MockClockTest.php:
```diff
@@ -0,0 +1,64 @@
+        $clock = new MockClock();
+        $this->assertSame('UTC', $clock->now()->getTimezone()->getName());
+
+        $tz = date_default_timezone_get();
+        $clock = new MockClock($tz);
+        $this->assertSame($tz, $clock->now()->getTimezone()->getName());
```
These tests unnecessarily depend on the global timezone settings. Let's decouple them.

```suggestion
        $clock = new MockClock('Europe/Paris');
        $this->assertSame('Europe/Paris', $clock->now()->getTimezone()->getName());
```

nicolas-grekas replied:
updated

derrabus reviewed (approved):
I like the component in its current state. Nice and simple. Let's see how it works for us when we dogfood it in other components.

GromNaN commented on src/Symfony/Component/Clock/SleepableClockInterface.php:
```diff
@@ -0,0 +1,22 @@
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+interface SleepableClockInterface
```
Nitpicking the interface and method name. `delay` and `DelayableClockInterface` would be semantically more correct. `sleep` is from a process pov, `delay` is about the clock.

nicolas-grekas replied:
I thought about "pause" also, but I kept sleep because to me this is the most expected and discoverable.
Any other name would introduce confusion IMHO, eg "delay": ":thinking: does that mean async?"

GromNaN replied:
My understanding may be biased. The method actually suspend the execution of the process; it will be understood as a wrapper for `sleep`. Ok for keeping this wording.

fabpot commented on src/Symfony/Component/Clock/README.md:
```diff
@@ -0,0 +1,45 @@
+Clock Component
+===============
+
+Symfony Clock enables decoupling applications from the system clock.
```
```suggestion
Symfony Clock decouples applications from the system clock.
```

andersonamuller commented on src/Symfony/Component/Clock/SleepableClockInterface.php:
```diff
@@ -0,0 +1,22 @@
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+interface SleepableClockInterface
```
```suggestion
interface TickingClockInterface
```

andersonamuller commented on src/Symfony/Component/Clock/SleepableClockInterface.php:
```diff
@@ -0,0 +1,22 @@
+{
+    public function now(): \DateTimeImmutable;
+
+    public function sleep(float|int $seconds): void;
```
```suggestion
    public function tick(float|int $seconds): void;
```

nicolas-grekas commented:
After reading https://github.com/php-fig/fig-standards/pull/1257 I'm wondering if it wouldn't make sense to add `withTimeZone()`?

That'd allow registering only one clock service and give a contract to users to stick to a TZ.

Loosely related, I'm wondering if we shouldn't swap our approach to naming and add `ClockInterface` to the component. That'd be the only interface provided by the component.

The benefit of this approach is that it would make us less dependant on the outcome and pace of the FIG.

My preference goes to doing both changes I propose here: one capable ClockInterface in the component and one more specific in PSR20 when it'll be released.

nicolas-grekas commented:
PR updated with the approach proposed in my previous comment:
The component now provides a `ClockInterface` with 3 methods: `now()`, `sleep()` and `withTimeZone()`.

I also registered a `clock` service into FrameworkBundle and a corresponding autowiring alias for ClockInterface.
Since we now have `withTimeZone()`, there is no need to provide several clock services (an UTC one and a default-TZ one.)


When PSR20 will be out (and if it remains compatible with the component), ppl will have the choice to use either the generic and narrow abstraction from the FIG - or the more capable one from the component when in need of extra features.

/cc @symfony/mergers even if you already voted, please vote again if you approve these changes.

yceruto reviewed (approved):
I loved it!

fabpot commented:
Thank you @nicolas-grekas.
